File: python/tests/manual_tester.py
""" Python SDK manual tester """

import getopt
import os
import sys
import time

if sys.path[0] != '':
    sys.path.insert(0, sys.path[0]+"/../")

from client import SyncServiceClient

def receive_callback(meta_data):
    parts = meta_data.object_id.split("@")
    if meta_data.deleted:
        os.remove("./" + parts[0])
        print "Deleted " + parts[0]
        sync_client.mark_object_deleted(meta_data)
    else:
        file = open("./" + parts[0], "w")
        if sync_client.fetch_object_data(meta_data, file):
            print "Received " + parts[0]
            sync_client.mark_object_consumed(meta_data)

def print_acls_helper(acl_type):
    if acl_type == "destination":
        acls, ok = sync_client.retrieve_all_destination_acls()    
    else:
        acls, ok = sync_client.retrieve_all_object_acls()
    if not ok:
        print "Failed to retrieve all", acl_type, "ACLs"
        return

    if len(acls) != 0:
        print acl_type, "ACLs:"
        
        for acl in acls:
            print "   ", acl + ":"

            if acl_type == "destination":
                usernames, ok = sync_client.retrieve_destination_acl(acl)    
            else:
                usernames, ok = sync_client.retrieve_object_acl(acl)
            if not ok:
                print "Failed to retrieve usernames from", acl_type, acl, "ACL"
                return

            print "       ", usernames
    else:
        print "There are no", acl_type, "ACLs"

def run_security_test():
    if len(dest_type) != 0:
        if remove:
            result = sync_client.remove_users_from_destination_acl(dest_type, [user])
            if not result:
                print "Failed to remove the user from the ACL"
        else:
            result = sync_client.add_users_to_destination_acl(dest_type, [user])
            if not result:
                print "Failed to add the user to the ACL"
    elif len(object_type) != 0:
        if remove:
            result = sync_client.remove_users_from_object_acl(object_type, [user])
            if not result:
                print "Failed to remove the user from the ACL"
        else:
            result = sync_client.add_users_to_object_acl(object_type, [user])
            if not result:
                print "Failed to add the user to the ACL"
    else:
        print_acls_helper("destination")
        print
        print_acls_helper("object")

protocol = "http"
server = "localhost:8080"
org_id = ""
object_type = ""
object_id = ""
dest_type = ""
dest_id = ""
remove = False
user = ""
cert = ""
key = ""
secret = ""
test = sys.argv[1].lower()

opts, args = getopt.getopt(sys.argv[2:], "p:s:", ["org=", "ot=", "oid=", "dt=", "id=", "remove", "user=", "cert=", "key=", "secret="])
for opt, value in opts:
    if opt == "-p":
        protocol = value
    elif opt == "-s":
        server = value
    elif opt == "--org":
        org_id = value
    elif opt == "--ot":
        object_type = value
    elif opt == "--oid":
        object_id = value
    elif opt == "--dt":
        dest_type = value
    elif opt == "--id":
        dest_id = value
    elif opt == "--remove":
        remove = True
    elif opt == "--user":
        user = value
    elif opt == "--cert":
        cert = value
    elif opt == "--key":
        key = value
    elif opt == "--secret":
        secret = value

if protocol != "unix" and protocol != "secure-unix":
    parts = server.split(":")
    if parts[0] != "":
        host = parts[0]
    if parts[1] != "":
        if parts[1].isdigit():
            port = int(parts[1])
        else:
            print "An invalid port number of ", parts[1], "was specified"
            sys.exit(99)
else:
    host = server
    port = 8080

sync_client = SyncServiceClient.Client(protocol, host, port)

if org_id != "":
    sync_client.set_org_id(org_id)
if cert != "":
    sync_client.set_ca_certificate(cert)
if key != "":
    sync_client.set_app_key_and_secret(key, secret)


if test == "deleteobject":
    if sync_client.delete_object(object_type, object_id):
        print "Object", object_type+":"+object_id, "deleted"
    else:
        print "Failed to delete object", object_type+":"+object_id

elif test == "deletewebhook":
    if sync_client.delete_webhook(object_type, "http://localhost:1234/webhook"):
        print "Webhook deleted"
    else:
        print "Failed to delete webhook"

elif test == "destinations":
    (results, ok) = sync_client.get_destinations()
    if ok:
        print "Retrieved destinations:"
        print results
    else:
        print "Failed to retrieve destinations"

elif test == "destinationobjects":
    (results, ok) = sync_client.get_destination_objects(dest_type, dest_id)
    if ok:
        print "Retrieved objects for destination:"
        print results
    else:
        print "Failed to retrieve objects for destination"

elif test == "object":
    (result, ok) = sync_client.get_object_metadata(object_type, object_id)
    if ok:
        if result != None:
            print "Retrieved object:"
            print result
        else:
            print "Object not found"
    else:
        print "Failed to retrieve object"

elif test == "objectdestinations":
    (results, ok) = sync_client.get_object_destinations(object_type, object_id)
    if ok:
        print "Retrieved destinations for object:"
        print results
    else:
        print "Failed to retrieve destinations for object"

elif test == "objectstatus":
    (result, ok) = sync_client.get_object_status(object_type, object_id)
    if ok:
        if result != None:
            print "Retrieved object status:", result
        else:
            print "Object not found"
    else:
        print "Failed to retrieve object status"

elif test == "registerwebhook":
    if sync_client.register_webhook(object_type, "http://localhost:1234/webhook"):
        print "Webhook registered"
    else:
        print "Failed to register webhook"

elif test == "resend":
    if sync_client.resend():
        print "Successfully request a resending of all objects"
    else:
        print "Failed to request a resending of all objects"

elif test == "security":
    run_security_test()

else:
    print "An invalid test of", os.argv[1], "was supplid."

File: src/main/java/com/horizon/syncservice/tests/ManualTester.java
package com.horizon.syncservice.tests;

import java.net.MalformedURLException;
import java.net.URL;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

import com.horizon.syncservice.client.SyncServiceClient;
import com.horizon.syncservice.client.SyncServiceMetaData;

import org.apache.commons.cli.CommandLine;
import org.apache.commons.cli.CommandLineParser;
import org.apache.commons.cli.DefaultParser;
import org.apache.commons.cli.HelpFormatter;
import org.apache.commons.cli.Options;
import org.apache.commons.cli.ParseException;

public class ManualTester {
    private enum Tests {
        HELP, OBJSTATUS, RESEND, SECURITY
    }

    private Tests test;
    private Map<Tests, Options> optionsMap = new HashMap<Tests, Options>();

    private String orgID;
    private String destType;
    private String destID;
    private String objectType;
    private String objectID;
    private String username;
    private boolean remove;
    private String serverProtocol;
    private String serverAddress;
    private String appKey;
    private String appSecret;

    public static void main(String[] argv) {
        ManualTester tester = new ManualTester();
        int rc = tester.parseArgs(argv);
        if (rc == 0) {
            rc = tester.runTests();
        }

        System.exit(rc);
    }

    private int runTests() {

        try {
            switch(test) {
                case HELP:
                    showUsage();
                    return 0;

                case OBJSTATUS:
                    return runObjectStatusTest();

                case RESEND:
                    return runResendTest();

                case SECURITY:
                    return runSecurityTest();
            }

            return 0;
        }
        catch (Exception e) {
            System.err.println("Failed to run test. Exception thrown: " + e.getMessage());
            e.printStackTrace();
            return 99;
        }
    }

    private int runObjectStatusTest() throws Exception {
        SyncServiceClient syncClient = createSyncServiceClient();
        String result = syncClient.getObjectStatus(objectType, objectID);
        System.out.println("The status of " + objectType + ":" + objectID + " is " + result);
        return 0;
    }

    private int runResendTest() throws Exception {
        SyncServiceClient syncClient = createSyncServiceClient();
        syncClient.resend();
        return 0;
    }

    private int runSecurityTest() throws Exception {
        SyncServiceClient syncClient = createSyncServiceClient();

        String[] usernames = new String[1];
        usernames[0] = username;
        if (destType.length() != 0) {
            if (remove) {
                syncClient.removeUsersFromDestinationACL(destType, usernames);
            }
            else {
                syncClient.addUsersToDestinationACL(destType, usernames);
            }
        }
        else if (objectType.length() != 0) {
            if (remove) {
                syncClient.removeUsersFromObjectACL(objectType, usernames);
            }
            else {
                syncClient.addUsersToObjectACL(objectType, usernames);
            }
        }
        else {
            printACLsHelper(syncClient, "destination");
            System.out.println();
            printACLsHelper(syncClient, "object");
        }
        return 0;
    }

    private void printACLsHelper(SyncServiceClient syncClient, String aclType) throws Exception {
        List<String> acls;
        if (aclType.equals("destination")) {
            acls = syncClient.retrieveAllDestinationACLs();
        }
        else {
            acls = syncClient.retrieveAllObjectACLs();
        }

        if (acls.size() > 0) {
            System.out.println(aclType + " ACLs:");
            for (String acl : acls) {
                List<String> usernames;
                if (aclType.equals("destination")) {
                    usernames = syncClient.retrieveDestinationACL(acl);
                }
                else {
                    usernames = syncClient.retrieveObjectACL(acl);
                }

                System.out.print("    " + acl + ":");
                for (int i=0  ;  i < usernames.size()  ;  i++) {
                    if (i % 8 == 0) {
                        System.out.print("\n        " + usernames.get(i));
                    }
                    else {
                        System.out.print(", " + usernames.get(i));
                    }
                }
                System.out.println();
            }
        }
        else {
            System.out.println("No " + aclType + " ACLs exist.");
        }
    }

    private int parseArgs(String[] args) {
        setupOptions();

        if (args.length < 1) {
            System.err.println("You must at least provide a test name");
            showUsage();
            return 1;
        }

        try {
            test = Tests.valueOf(args[0].toUpperCase());
        }
        catch(IllegalArgumentException e) {
            System.err.println("The first argument " + args[0] + " is invalid.");
            showUsage();
            return 1;
        }

        Options options = optionsMap.get(test);
        if (options == null) {
            return 0;
        }

        CommandLineParser parser = new DefaultParser();
        try {
            CommandLine cmd = parser.parse(options, args);

            orgID = cmd.getOptionValue("orgID", "");
            destType = cmd.getOptionValue("destType", "");
            destID = cmd.getOptionValue("destID", "");
            objectType = cmd.getOptionValue("objectType", "");
            objectID = cmd.getOptionValue("objectID", "");
            username = cmd.getOptionValue("username", "");
            remove = cmd.hasOption("remove");

            serverProtocol = cmd.getOptionValue("serverProtocol", "https");
            serverAddress = cmd.getOptionValue("serverAddress", "localhost:8443");

            appKey = cmd.getOptionValue("appKey", "");
            appSecret = cmd.getOptionValue("appSecret", "");
        }
        catch (ParseException e) {
            System.out.println(e.getMessage());

            return 1;
        }

        return 0;
    }

    private void setupOptions() {
        Options options;

        // objectStatus
        options = new Options();
        addCommonOptions(options);
        options.addOption("org", "orgID", true, "Specify the organization ID of the object (optional)");
        options.addOption("ot", "objectType", true, "Specify the object type of the object");
        options.addOption("oid", "objectID", true, "Specify the object ID of the objects");
        optionsMap.put(Tests.OBJSTATUS, options);

        // resend
        options = new Options();
        addCommonOptions(options);
        optionsMap.put(Tests.RESEND, options);

        // security
        options = new Options();
        addCommonOptions(options);
        options.addOption("org", "orgID", true, "Specify the organization ID of the ACL (optional)");
        options.addOption("dt", "destType", true, "Specify the destination type of the ACL");
        options.addOption("ot", "objectType", true, "Specify the object type of the ACL");
        options.addOption("id", "username", true, "Specify the user name to be added/removed from the ACL");
        options.addOption("remove", "remove", false, "Indicated that the user shoul be removed from the ACL");
        optionsMap.put(Tests.SECURITY, options);
    }

    private void addCommonOptions(Options options) {
        options.addOption("p", "serverProtocol", true, "Specify the protocol of the Cloud Sync Service");
        options.addOption("s", "serverAddress", true, "Specify the address and port of the Cloud Sync Service");
        options.addOption("key", "appKey", true, "Specify the app key to be used when connecting to the Sync Service");
        options.addOption("secret", "appSecret", false, "Specify the app secret to be used when connecting to the Sync Service");
    }

    private SyncServiceClient createSyncServiceClient() throws MalformedURLException {
        return new SyncServiceClient.Builder()
           .withUrl(new URL(serverProtocol + "://" + serverAddress))
           .withAppKeyAndAppSecret(appKey, appSecret)
           .withOrgID(orgID)
           .build();
    }

    private void showUsage() {
        System.out.println("ManualTester help");
        System.out.println("");
        System.out.println("ManualTester objstatus");
        System.out.println("");
        System.out.println("ManualTester resend");
        System.out.println("");
        System.out.println("ManualTester security");
    }
}